src/lab4.py

```diff
@@ -54,124 +54,125 @@ def request(url):
     return headers, body
 
 class Text:
-    def __init__(self, text):
+    def __init__(self, text, parent):
         self.text = text
+        self.children = []
+        self.parent = parent
 
     def __repr__(self):
-        return "\"" + self.text.replace("\n", "\\n") + "\""
+        return repr(self.text)
 
-SELF_CLOSING_TAGS = [
-    "area", "base", "br", "col", "embed", "hr", "img", "input",
-    "link", "meta", "param", "source", "track", "wbr",
-]
+class Element:
+    def __init__(self, tag, attributes, parent):
+        self.tag = tag
+        self.attributes = attributes
+        self.children = []
+        self.parent = parent
 
-class Tag:
-    def __init__(self, text):
+    def __repr__(self):
+        attrs = [k + "=\"" + v + "\"" for k, v  in self.attributes.items()]
+        return "<" + self.tag + " " + " ".join(attrs) + ">"
```

Comment:
Optimize to avoid the extra space before the > or between / and >; looks ugly in the printed output and is not what real browsers or xml printers ever show.